### DISCUSSION TITLE
Migrating from ActiveMQ to EMQX

### DISCUSSION THREAD

**Discussion by pkirk** [emqx/emqx#12563] [Q&A]
Dear Community,
I would like to swap an ActiveMQ with an EMQX, with an already existing base of IoT clients that simply use nodeJS  in a similiar way like the one here below:

```
const TRUSTED_CA_LIST = fs.readFileSync(path.join(__dirname, "/certs/amq-server_cert.pem"));
const connectOptions = {
    rejectUnauthorized: false,
    ca: TRUSTED_CA_LIST,
};
const client = mqtt.connect(`mqtts://${process.env.MQTT_HOST}:${process.env.MQTT_PORT}`, connectOptions);
```

The ActiveMQ has this two certs configurations:
jetty.xml:			<property name="keyStorePath" value="${activemq.conf}/broker.ks" />
activemq.xml:		<sslContext keyStore="file:/opt/activemq/conf/ssl-certs/amq-server.ks"
(with their relative password)

I converted those two KeyStore with the following commands
$ keytool -importkeystore -srckeystore your_keystore.ks -destkeystore your_keystore.p12 -deststoretype PKCS12
$ openssl pkcs12 -in your_keystore.p12 -out your_keystore.pem

Inside the broker.pem there was the certificate and the key, while in amq-server.ks there was only the certificate, and it was different from the one found in the IoT Node client amq-server_cert.pem.

But what and where do I load them on the EMQX ssl listener?
It needs a TLS Cert, a TLS Key and an optional CA Cert.

I tried to put the brokers' but the connection won't go, I see a
`ERROR CONNECTED Error: Client network socket disconnected before secure TLS connection was established`
on the client and a 
`2024-02-21T21:32:57.886759+01:00 [notice] supervisor: {esockd_connection_sup,<0.3184.0>}, errorContext: ssl_error, reason: {options,incompatible,[{verify,verify_peer},{cacerts,undefined}]}, offender: [{pid,<0.3184.0>},{name,connection},{mfargs,{emqx_connection,start_link,[#{listener => {ssl,default},limiter => undefined,zone => default,enable_authn => true}]}}]
`
on the EMQX.
If I disactivate the TLS Verify Client, I get
`2024-02-21T21:43:33.476893+01:00 [error] supervisor: {esockd_connection_sup,<0.3383.0>}, errorContext: connection_shutdown, reason: {ssl_failure,{{{case_clause,{error,{asn1,{bad_range,[{'OTP-PUB-KEY','dec_OTP-X520countryname',2,[{file,"../src/OTP-PUB-KEY.erl"},{line,18031}]},{'OTP-PUB-KEY',decode,2,[{file,"../src/OTP-PUB-KEY.erl"},{line,1239}]},{pubkey_cert_records,transform,2,[{file,"pubkey_cert_records.erl"},{line,65}]},{lists,map,2,[{file,"lists.erl"},{line,1559}]},{pubkey_cert_records,transform,2,[{file,"pubkey_cert_records.erl"},{line,92}]},{pubkey_cert_records,decode_tbs,1,[{file,"pubkey_cert_records.erl"},{line,330}]},{pubkey_cert_records,decode_cert,1,[{file,"pubkey_cert_records.erl"},{line,42}]},{public_key,pkix_decode_cert,2,[{file,"public_key.erl"},{line,529}]},{ssl_handshake,get_cert_params,1,[{file,"ssl_handshake.erl"},{line,1791}]},{tls_handshake_1_3,get_certificate_params,1,[{file,"tls_handshake_1_3.erl"},{line,1552}]},{tls_server_connection_1_3,select_server_cert_key_pair,7,[{file,"tls_server_connection_1_3.erl"},{line,694}]},{tls_server_connection_1_3,do_handle_client_hello,2,[{file,"tls_server_connection_1_3.erl"},{line,440}]},{tls_server_connection_1_3,handle_client_hello,2,[{file,"tls_server_connection_1_3.erl"},{line,380}]},{gen_statem,loop_state_callback,11,[{file,"gen_statem.erl"},{line,1395}]},{tls_connection,init,1,[{file,"tls_connection.erl"},{line,162}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,241}]}]}}}},[{pubkey_cert_records,transform,2,[{file,"pubkey_cert_records.erl"},{line,65}]},{lists,map,2,[{file,"lists.erl"},{line,1559}]},{pubkey_cert_records,transform,2,[{file,"pubkey_cert_records.erl"},{line,92}]},{pubkey_cert_records,decode_tbs,1,[{file,"pubkey_cert_records.erl"},{line,330}]},{pubkey_cert_records,decode_cert,1,[{file,"pubkey_cert_records.erl"},{line,42}]},{public_key,pkix_decode_cert,2,[{file,"public_key.erl"},{line,529}]},{ssl_handshake,get_cert_params,1,[{file,"ssl_handshake.erl"},{line,1791}]},{tls_handshake_1_3,get_certificate_params,1,[{file,"tls_handshake_1_3.erl"},{line,1552}]},{tls_server_connection_1_3,select_server_cert_key_pair,7,[{file,"tls_server_connection_1_3.erl"},{line,694}]},{tls_server_connection_1_3,do_handle_client_hello,2,[{file,"tls_server_connection_1_3.erl"},{line,440}]},{tls_server_connection_1_3,handle_client_hello,2,[{file,"tls_server_connection_1_3.erl"},{line,380}]},{gen_statem,loop_state_callback,11,[{file,"gen_statem.erl"},{line,1395}]},{tls_connection,init,1,[{file,"tls_connection.erl"},{line,162}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,241}]}]},{gen_statem,call,[<0.3388.0>,{start,15000},infinity]}}}, offender: [{pid,<0.3383.0>},{name,connection},{mfargs,{emqx_connection,start_link,[#{listener => {ssl,default},limiter => undefined,zone => default,enable_authn => true}]}}]`

Can anyone please help and advice? Thanks for your time!

**Comment by thalesmg** [comment]
Hi!

If I understood correctly, your setup only depends on how your client certificates were issued and if you want to verify the client certificates during connection.

It sounds like this `/opt/activemq/conf/ssl-certs/amq-server.ks` file you are trying to use appears to contain the _server_ certificates and keys that ActiveMQ server uses.  You could alternatively issue new certificates for EMQX and just set them up via the dashboard UI, simply by uploading them.  For reference: https://docs.emqx.com/en/enterprise/latest/network/emqx-mqtt-tls.html#enable-ssl-tls-with-one-way-authentication

I see your clients are already using a server certificate to authenticate (`/certs/amq-server_cert.pem`).  So, if you use EMQX's certificate, it's a matter of using the new file rather than this one.

Otherwise, if you want to use the exact same server certificate as before, you'll need to extract both the certificate and intermediate/root CA from the keystore.  Perhaps this could be a starting point for extracting the CA from there: https://stackoverflow.com/a/46371292/2708711

**Reply by pkirk** [reply]
@thalesmg I tried that extraction but the problem seems to be the one pointed by @zmstone , the certificate decoding.

**Comment by zmstone** [comment]
It seems that the server failed to decode client certificate.
More specifically, it's the country name `C` part of DN, i.e. in x509 certificate Subject.

Erlang SSL lib only allows 2 or 3 characters for country name.
e.g. `C=se, ` for Sweden etc.

And Openssl command line seems to only allow exactly 2 characters.

**Reply by thalesmg** [reply]
@pkirk Could you share the certificates used in the case where you disabled peer verification (the one with the larger log) so we may inspect the DN?  Both the client and the server certificates (`/certs/amq-server_cert.pem`) would be helpful.

**Reply by pkirk** [reply]
@thalesmg  I would prefer not to, because it's a sensitive information, that IoT is inside a security network.

In the meantime I swapped the cert on the client with a new one (the one that EMQX generates upon installation) and it works.

But I don't think I can find an easy solution, without patching EMQX to accept a malformed client certificate.

Thanks!

**Reply by zmstone** [reply]
Hi @pkirk 
I'm afraid there is no easy way to patch EMQX to accept malformed certificates.
It's the underlying SSL library as a part of Erlang/OTP which has a strict parser for certificates.

The certificate is arguably not sensitive, but since you can confirm that other certificates work fine,
I don't think it's actually necessary to share the cert with us.
